onnxruntime/core/providers/cuda/tensor/tile_impl.cu

```diff
@@ -7,140 +7,178 @@
 namespace onnxruntime {
 namespace cuda {
 
-constexpr int MAX_DIMS = 16;
+#ifdef USE_ROCM
+constexpr int num_elements_per_thread = 2;
+constexpr int num_threads_per_block = 512;
+#else
+constexpr int num_elements_per_thread = GridDim::maxElementsPerThread;
+constexpr int num_threads_per_block = GridDim::maxThreadsPerBlock;
+#endif
 
 template <typename T>
-__global__ void _UnRolledTileKernel(
-    const size_t shape_rank,
-    const TArray<fast_divmod> fdm_input_shape,
-    const TArray<int64_t> input_strides,
-    const T* input_data,
-    const TArray<fast_divmod> fdm_output_strides,
-    T* output_data,
-    const CUDA_LONG N) {
-  CALCULATE_ELEMENTWISE_INDEX_OR_EXIT(id, N);
-  CUDA_LONG input_index = 0;
-  CUDA_LONG output_index = id;
+__global__ void _UnRolledTileKernel(const size_t shape_rank, const TArray<fast_divmod> fdm_input_shape,
+                                    const TArray<int64_t> input_strides, const T* input_data,
+                                    const TArray<fast_divmod> fdm_output_strides, T* output_data, const CUDA_LONG N) {
+  CUDA_LONG start = num_elements_per_thread * num_threads_per_block * blockIdx.x + threadIdx.x;
+  T value[num_elements_per_thread];
+  CUDA_LONG id = start;
+#pragma unroll
+  for (int i = 0; i < num_elements_per_thread; ++i) {
+    if (id < N) {
+      CUDA_LONG input_index = 0;
+      CUDA_LONG offset = id;
+#pragma unroll
+      for (auto dim = 0; dim < fdm_output_strides.Capacity(); ++dim) {
```

Comment:
nit: Call fdm_output_strides.Capacity() once and cache it ?

Reply:
My understanding is capacity is the tempate value, which is a constexpr, and it's >= TArray size. So actually during the compile it's a constexpr. I.E., the call is handled during compile, not the real execution.

static constexpr int32_t Capacity() { return capacity; };